Duplicate key exceptions in diagram layout handling can render a diagram unusable

I don't have (yet) exact reproduction steps, but the general scenario is to have a relatively large project (e.g. the "Sirius Web" Papaya example) and a unsynchronized diagram opened (it can be almost empty). When creating several new nodes in rapid succession by DnD from the explorer, at some point on the frontend I get errors:

![image](https://github.com/eclipse-sirius/sirius-web/assets/10608/0af21359-e280-41cc-9e71-390c3955e342)

On the backend I see several stack traces with messages like:

```
Caused by: java.lang.IllegalStateException: Duplicate key 5193ff79-8b9a-3654-94a7-46add7d92364 (attempted merging values NodeLayoutData[id=5193ff79-8b9a-3654-94a7-46add7d92364, position=Position[x=-4763.55109521028, y=654.013054906542], size=Size[width=378.78125, height=770.0], resizedByUser=false] and NodeLayoutData[id=5193ff79-8b9a-3654-94a7-46add7d92364, position=Position[x=-4763.55109521028, y=654.013054906542], size=Size[width=378.78125, height=770.0], resizedByUser=false])
	at java.base/java.util.stream.Collectors.duplicateKeyException(Collectors.java:135) ~[na:na]
	at java.base/java.util.stream.Collectors.lambda$uniqKeysMapAccumulator$1(Collectors.java:182) ~[na:na]
	at java.base/java.util.stream.ReduceOps$3ReducingSink.accept(ReduceOps.java:169) ~[na:na]
	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1625) ~[na:na]
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509) ~[na:na]
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499) ~[na:na]
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:921) ~[na:na]
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234) ~[na:na]
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:682) ~[na:na]
	at org.eclipse.sirius.components.collaborative.diagrams.DiagramEventProcessor.handle(DiagramEventProcessor.java:138) ~[classes/:na]
```

I'm not sure exactly why we get duplicate keys here, but once they occur, the 2-args version of `Collectors.toMap()` we use throws an exception in the middle of the `DiagramEventProcessor.handle` execution, which is not handled and causes all kinds of issues.

I'll continue to invesitgate the root cause of the duplicates, but also propose an PR to use the 3-args version of `Collectors.toMap()` in the few places where we use the 2-args one to at least mitigate the consequences.

--- a/CHANGELOG.adoc
+++ b/CHANGELOG.adoc
@@ -40,7 +40,7 @@ More existing APIs will be migrated to this new common pattern.
 - [releng] Remove the dependency to `reflect-metadata`
 - [releng] Switch to `subscriptions-transport-ws` 0.11.0
 - https://github.com/eclipse-sirius/sirius-web/issues/3466[#3466] [gantt] Move to @ObeoNetwork/gantt-task-react 0.4.14 to benefit for enhancements
-- [releng] Switch to EMF Json 2.3.10
+- [releng] Switch to EMF Json 2.3.11
 
 
 === Bug fixes

--- a/CHANGELOG.adoc
+++ b/CHANGELOG.adoc
@@ -56,6 +56,7 @@ More existing APIs will be migrated to this new common pattern.
 - https://github.com/eclipse-sirius/sirius-web/issues/3582[#3582] [tree] Restore the initial direct edit tree item label for the explorer
 - https://github.com/eclipse-sirius/sirius-web/issues/3611[#3611] [diagram] Fix missing creation tool image in the contextual palette
 - https://github.com/eclipse-sirius/sirius-web/issues/3628[#3628] [sirius-web] Restore support for expand all and reveal in the explorer
+- https://github.com/eclipse-sirius/sirius-web/issues/3616[#3616] [diagram] Fix potential exceptions due to duplicate keys in diagram event processing
 
 === New Features
 

--- a/packages/diagrams/backend/sirius-components-collaborative-diagrams/src/main/java/org/eclipse/sirius/components/collaborative/diagrams/DiagramEventProcessor.java
+++ b/packages/diagrams/backend/sirius-components-collaborative-diagrams/src/main/java/org/eclipse/sirius/components/collaborative/diagrams/DiagramEventProcessor.java
@@ -137,7 +137,8 @@ public class DiagramEventProcessor implements IDiagramEventProcessor {
                 var nodeLayoutData = layoutDiagramInput.diagramLayoutData().nodeLayoutData().stream()
                         .collect(Collectors.toMap(
                                 NodeLayoutDataInput::id,
-                                nodeLayoutDataInput -> new NodeLayoutData(nodeLayoutDataInput.id(), nodeLayoutDataInput.position(), nodeLayoutDataInput.size(), nodeLayoutDataInput.resizedByUser())
+                                nodeLayoutDataInput -> new NodeLayoutData(nodeLayoutDataInput.id(), nodeLayoutDataInput.position(), nodeLayoutDataInput.size(), nodeLayoutDataInput.resizedByUser()),
+                                (oldValue, newValue) -> newValue
                         ));
 
                 var layoutData = new DiagramLayoutData(nodeLayoutData, Map.of(), Map.of());

--- a/packages/diagrams/backend/sirius-components-diagrams/src/main/java/org/eclipse/sirius/components/diagrams/components/EdgesRequestor.java
+++ b/packages/diagrams/backend/sirius-components-diagrams/src/main/java/org/eclipse/sirius/components/diagrams/components/EdgesRequestor.java
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2019, 2021 Obeo.
+ * Copyright (c) 2019, 2024 Obeo.
  * This program and the accompanying materials
  * are made available under the terms of the Eclipse Public License v2.0
  * which accompanies this distribution, and is available at
@@ -30,10 +30,8 @@ public class EdgesRequestor implements IEdgesRequestor {
     private final Map<String, Edge> edgeId2Edges;
 
     public EdgesRequestor(List<Edge> previousEdges) {
-        // @formatter:off
         this.edgeId2Edges = previousEdges.stream()
-                .collect(Collectors.toMap(Edge::getId, Function.identity()));
-        // @formatter:on
+                .collect(Collectors.toMap(Edge::getId, Function.identity(), (oldValue, newValue) -> newValue));
     }
 
     @Override

--- a/packages/diagrams/backend/sirius-components-diagrams/src/main/java/org/eclipse/sirius/components/diagrams/components/NodeComponent.java
+++ b/packages/diagrams/backend/sirius-components-diagrams/src/main/java/org/eclipse/sirius/components/diagrams/components/NodeComponent.java
@@ -55,7 +55,7 @@ import org.eclipse.sirius.components.representations.VariableManager;
 public class NodeComponent implements IComponent {
 
     public static final String COLLAPSING_STATE = "collapsingState";
-    
+
     public static final String IS_BORDER_NODE = "isBorderNode";
 
     public static final String SEMANTIC_ELEMENT_IDS = "semanticElementIds";
@@ -83,9 +83,9 @@ public class NodeComponent implements IComponent {
                     .map(ViewCreationRequest::getTargetObjectId)
                     .toList();
             List<String> previousNodeIds = this.props.getPreviousTargetObjectIds();
-            List<String> semanticElementIds = new ArrayList<>(creationRequestsIds);
+            Set<String> semanticElementIds = new HashSet<>(creationRequestsIds);
             semanticElementIds.addAll(previousNodeIds);
-            nodeComponentVariableManager.put(SEMANTIC_ELEMENT_IDS, semanticElementIds);
+            nodeComponentVariableManager.put(SEMANTIC_ELEMENT_IDS, semanticElementIds.stream().toList());
         }
 
         List<Element> children = new ArrayList<>();

--- a/packages/diagrams/backend/sirius-components-diagrams/src/main/java/org/eclipse/sirius/components/diagrams/components/NodesRequestor.java
+++ b/packages/diagrams/backend/sirius-components-diagrams/src/main/java/org/eclipse/sirius/components/diagrams/components/NodesRequestor.java
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2019, 2020 Obeo.
+ * Copyright (c) 2019, 2024 Obeo.
  * This program and the accompanying materials
  * are made available under the terms of the Eclipse Public License v2.0
  * which accompanies this distribution, and is available at
@@ -30,10 +30,8 @@ public class NodesRequestor implements INodesRequestor {
     private final Map<String, Node> targetObjectId2Nodes;
 
     public NodesRequestor(List<Node> previousNodes) {
-        // @formatter:off
         this.targetObjectId2Nodes = previousNodes.stream()
-                .collect(Collectors.toMap(Node::getTargetObjectId, Function.identity()));
-        // @formatter:on
+                .collect(Collectors.toMap(Node::getTargetObjectId, Function.identity(), (oldValue, newValue) -> newValue));
     }
 
     @Override

--- a/packages/emf/backend/sirius-components-emf/pom.xml
+++ b/packages/emf/backend/sirius-components-emf/pom.xml
@@ -102,7 +102,7 @@
 		<dependency>
 			<groupId>org.eclipse.sirius.emfjson</groupId>
 			<artifactId>org.eclipse.sirius.emfjson</artifactId>
-			<version>2.3.10-SNAPSHOT</version>
+			<version>2.3.11-SNAPSHOT</version>
 		</dependency>
 		<dependency>
 			<groupId>org.springframework.boot</groupId>

